staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+	"runtime"
+
+	// TODO: decide whether to also generate the old metrics, which
+	// categorize according to mutating vs readonly.
+
+	// "k8s.io/apiserver/pkg/endpoints/metrics"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
+	utilfilterconfig "k8s.io/apiserver/pkg/util/flowcontrol/filterconfig"
+	"k8s.io/klog"
+)
+
+// WithPriorityAndFairness limits the number of in-flight
+// requests in a fine-grained way.
+func WithPriorityAndFairness(
+	handler http.Handler,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+	fcIfc utilflowcontrol.Interface,
+) http.Handler {
+	if fcIfc == nil {
+		klog.Warningf("priority and fairness support not found, skipping")
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no RequestInfo found in context"))
+			return
+		}
+		user, ok := apirequest.UserFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no User found in context"))
+			return
+		}
+		requestDigest := utilfilterconfig.RequestDigest{requestInfo, user}
+
+		// Skip tracking long running requests.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			klog.V(6).Infof("Serving RequestInfo=%#+v, user.Info=%#+v as longrunning\n", requestInfo, user)
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		execute, afterExecute := fcIfc.Wait(ctx, requestDigest)
+		if !execute {
+			tooManyRequests(r, w)
+			return
+		}
+		defer afterExecute()
```

Comment:
wrt @lavalamp's previous suggestion about attaching the name of the matched FS/PL to the response header, it doesnt look easy to get know which FS/PL matched from this encapsulation. probably we can refactor `afterExecute` to return a `ResponseDigest` storing the information so that we attach the header on this defer call

Reply:
I think we agreed to address that in a follow-up PR.

Reply:
yup.